File: packages/javascript/src/subscribers/graceful-shutdown.ts
import { Consumer } from './consumer';

/**
 * Logger interface for graceful shutdown
 */
interface Logger {
  info(message: string, metadata?: Record<string, unknown>): void;
  warn(message: string, metadata?: Record<string, unknown>): void;
  error(message: string, metadata?: Record<string, unknown>): void;
  debug(message: string, metadata?: Record<string, unknown>): void;
}

/**
 * Graceful shutdown configuration
 */
export interface GracefulShutdownOptions {
  timeout?: number; // milliseconds
  logger?: Logger;
}

/**
 * Graceful shutdown manager for subscribers
 * Handles signal trapping and ensures clean shutdown with message drain.
 *
 * Uses Consumer.stop() which calls NATS connection.drain() under the hood.
 * NATS drain stops accepting new messages, waits for in-flight messages to
 * complete, then closes the connection — providing built-in graceful shutdown.
 */
export class GracefulShutdown {
  private readonly timeout: number;
  private readonly logger?: Logger;
  private shuttingDown = false;

  private static readonly DEFAULT_TIMEOUT = 30_000; // 30 seconds

  constructor(
    private readonly consumer: Consumer,
    options: GracefulShutdownOptions = {}
  ) {
    this.timeout = options.timeout || GracefulShutdown.DEFAULT_TIMEOUT;
    this.logger = options.logger;
  }

  /**
   * Start graceful shutdown process.
   * Calls consumer.stop() with a timeout to ensure shutdown completes.
   * NATS drain handles waiting for in-flight messages.
   * @returns True if shutdown completed within timeout
   */
  async shutdown(): Promise<boolean> {
    if (this.shuttingDown) {
      return false;
    }

    this.shuttingDown = true;
    const startedAt = Date.now();

    this.logger?.info('Starting graceful shutdown', { timeout: this.timeout });

    try {
      // consumer.stop() calls connection.drain() + connection.close()
      // which handles the graceful message drain natively
      await Promise.race([
        this.consumer.stop(),
        this.sleep(this.timeout).then(() => {
          throw new Error('Shutdown timeout reached');
        }),
      ]);

      const elapsed = Date.now() - startedAt;
      this.logger?.info('Graceful shutdown complete', { elapsed, graceful: true });
      return true;
    } catch (error) {
      const elapsed = Date.now() - startedAt;
      this.logger?.warn('Graceful shutdown timed out, forcing stop', {
        elapsed,
        timeout: this.timeout,
      });

      // Force stop by calling stop again (idempotent)
      try {
        await this.consumer.stop();
      } catch {
        // Already stopped or connection lost — safe to ignore
      }

      return false;
    }
  }

  /**
   * Check if shutdown is in progress
   */
  isShuttingDown(): boolean {
    return this.shuttingDown;
  }

  /**
   * Install signal handlers for graceful shutdown
   * Traps SIGTERM and SIGINT
   */
  installSignalHandlers(): void {
    const signals: NodeJS.Signals[] = ['SIGTERM', 'SIGINT'];

    signals.forEach((signal) => {
      process.on(signal, () => {
        this.logger?.info(`Received ${signal}, initiating shutdown`);
        this.shutdown()
          .then(() => process.exit(0))
          .catch((error) => {
            this.logger?.error('Shutdown failed', { error });
            process.exit(1);
          });
      });
    });
  }

  private sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
  }
}

File: packages/ruby/lib/nats_pubsub/subscribers/graceful_shutdown.rb
# frozen_string_literal: true

module NatsPubsub
  module Subscribers
    # Graceful shutdown manager for subscribers
    # Handles signal trapping and ensures clean shutdown with message drain
    class GracefulShutdown
      DEFAULT_TIMEOUT = 30 # seconds

      attr_reader :worker, :timeout, :logger

      # Initialize graceful shutdown manager
      #
      # @param worker [Worker] Worker instance to manage
      # @param timeout [Integer] Shutdown timeout in seconds
      # @param logger [Logger] Logger instance
      def initialize(worker, timeout: DEFAULT_TIMEOUT, logger: nil)
        @worker = worker
        @timeout = timeout
        @logger = logger || NatsPubsub.config.logger
        @shutting_down = false
        @shutdown_started_at = nil
      end

      # Start graceful shutdown process
      #
      # @return [Boolean] True if shutdown completed gracefully within timeout
      def shutdown
        return false if @shutting_down

        @shutting_down = true
        @shutdown_started_at = Time.now

        logger&.info('Starting graceful shutdown', timeout: timeout)

        # Stop accepting new messages
        stop_accepting_messages

        # Wait for in-flight messages
        completed = wait_for_completion

        # Force terminate if needed
        force_terminate unless completed

        # Close connections
        close_connections

        elapsed = (Time.now - @shutdown_started_at).round(1)
        logger&.info('Graceful shutdown complete', elapsed: elapsed, graceful: completed)

        completed
      end

      # Check if shutdown is in progress
      #
      # @return [Boolean] True if shutdown is in progress
      def shutting_down?
        @shutting_down
      end

      # Install signal handlers for graceful shutdown
      # Traps SIGTERM and SIGINT
      def install_signal_handlers
        %w[TERM INT].each do |signal|
          Signal.trap(signal) do
            logger&.info("Received SIG#{signal}, initiating shutdown")

            begin
              shutdown
              exit(0)
            rescue StandardError => e
              logger&.error('Shutdown failed', error: e.message)
              exit(1)
            end
          end
        end
      end

      private

      # Stop accepting new messages
      def stop_accepting_messages
        logger&.info('Stopping message acceptance')
        worker.pause! if worker.respond_to?(:pause!)
      end

      # Wait for in-flight messages to complete
      #
      # @return [Boolean] True if all messages completed within timeout
      def wait_for_completion
        deadline = Time.now + timeout

        loop do
          in_flight = worker.in_flight_count

          if in_flight.zero?
            logger&.info('All messages processed')
            return true
          end

          if Time.now >= deadline
            logger&.warn('Shutdown timeout reached',
                         in_flight: in_flight,
                         timeout: timeout)
            return false
          end

          logger&.debug('Waiting for messages',
                        in_flight: in_flight,
                        remaining: (deadline - Time.now).round(1))
          sleep 0.5
        end
      end

      # Force terminate remaining messages
      def force_terminate
        logger&.warn('Force terminating remaining messages')
        worker.force_stop! if worker.respond_to?(:force_stop!)
      end

      # Close connections
      def close_connections
        logger&.info('Closing connections')
        worker.close if worker.respond_to?(:close)
      end
    end
  end
end
